spec/datadog/appsec/configuration/settings_spec.rb

```diff
@@ -0,0 +1,192 @@
+    end
+  end
+
+  describe Datadog::AppSec::Configuration::Settings do
+    include_context 'registry with integration'
+
+    subject(:settings) { described_class.new }
+
+    let(:dsl) { Datadog::AppSec::Configuration::DSL.new }
+
+    after { settings.send(:reset!) }
+
+    describe '#enabled' do
+      subject(:enabled) { settings.enabled }
+      it { is_expected.to eq(true) }
+    end
+
+    describe '#enabled=' do
+      subject(:enabled_) { settings.merge(dsl.tap { |c| c.enabled = false }) }
+      it { expect { enabled_ }.to change { settings.enabled }.from(true).to(false) }
+    end
+
+    describe '#ruleset' do
+      subject(:ruleset) { settings.ruleset }
+      it { is_expected.to eq(:recommended) }
+    end
+
+    describe '#ruleset=' do
+      subject(:ruleset_) { settings.merge(dsl.tap { |c| c.ruleset = :risky }) }
+      it { expect { ruleset_ }.to change { settings.ruleset }.from(:recommended).to(:risky) }
+    end
+
+    describe '#waf_timeout' do
+      subject(:waf_timeout) { settings.waf_timeout }
+      it { is_expected.to eq(5000) }
+    end
+
+    describe '#waf_timeout=' do
+      subject(:waf_timeout_) { settings.merge(dsl.tap { |c| c.waf_timeout = 3 }) }
+      it { expect { waf_timeout_ }.to change { settings.waf_timeout }.from(5000).to(3) }
+    end
+
+    describe '#waf_debug' do
+      subject(:waf_debug) { settings.waf_debug }
+      it { is_expected.to eq(false) }
+    end
+
+    describe '#waf_debug=' do
+      subject(:waf_debug_) { settings.merge(dsl.tap { |c| c.waf_debug = true }) }
+      it { expect { waf_debug_ }.to change { settings.waf_debug }.from(false).to(true) }
+    end
+
+    describe '#trace_rate_limit' do
+      subject(:trace_rate_limit) { settings.trace_rate_limit }
+      it { is_expected.to eq(100) }
+    end
+
+    describe '#trace_rate_limit=' do
+      subject(:trace_rate_limit_) { settings.merge(dsl.tap { |c| c.trace_rate_limit = 2 }) }
+      it { expect { trace_rate_limit_ }.to change { settings.trace_rate_limit }.from(100).to(2) }
+    end
+
+    describe '#[]' do
+      describe 'when the integration exists' do
+        subject(:get) { settings[integration_name] }
+
+        let(:integration_options) { { foo: :bar } }
+
+        before { settings.merge(dsl.tap { |c| c.instrument(integration_name, integration_options) }) }
+
+        it 'retrieves the described configuration' do
+          is_expected.to eq(integration_options)
+        end
+      end
+
+      context 'when the integration doesn\'t exist' do
+        it do
+          expect { settings[:foobar] }.to raise_error(ArgumentError, /foobar/)
+        end
+      end
+    end
+
+    context 'with env vars' do
+      before do
+        stub_const('ENV', {})
+        allow(ENV).to receive(:[]).and_call_original
+      end
+
+      describe 'DD_APPSEC_ENABLED' do
+        before do
+          allow(ENV).to receive(:[]).with('DD_APPSEC_ENABLED').and_return('1')
+        end
```

Comment:
Oh, I did not know that and found references to a similar approach in the specs, so did not look further.

TBH I'm not too sure of the added value of ClimateControl (is that mainly about [this `synchronize`](https://github.com/thoughtbot/climate_control/blob/4cf92235b66170504e05511205fd5c0d8e8a0a01/lib/climate_control/modifier.rb#L10)? Are tests ever run multithreaded?) and kind of find it less readable as well as being unable to assert whether `ENV` has received `:[]` for a given key.

Reply:
One thing I think is an advantage is when debugging tests. When I was looking at the `DD_APPSEC_ENABLED` tests I was getting pretty confused where the env var was coming from because I was printing `ENV` and did not see it there, but `ENV['DD_APPSEC_ENABLED']` returned a value.

But... I don't feel too strongly, just thought I'd mention it :)